### ISSUE TITLE
spacings in hstack/vstack/gridstack

### ISSUE THREAD

**Issue by IgorDouven** [GiovineItalia/Gadfly.jl#1169]
Is there a way to fine-tune spacings when using `hstack`/`vstack`/`gridstack`? I tried to control this via `plot_padding` for the individual plots that are being stacked, but this doesn't allow me to reduce vertical space between the rows (which is badly needed in my plot). Something like `Spacings` in Mathematica's `GraphicsGrid` would be nice to have.

**Comment by IgorDouven** [comment]
Thanks, but this didn't work for me. I would now have to first use `hstack` to collect some plots into row and then I would want to use `vstack2` to collate those rows. However, in that case the input for the latter function is of the wrong type (`Compose.Context` instead of `Vector{Plot}`). Is there a way around this?

**Comment by IgorDouven** [comment]
This worked for me as well -- great many thanks! I'm closing the issue now, though it would be nice to have this function in the documentation.

**Comment by bjarthur** [comment]
let's leave this open as adding this functionality would be useful

**Comment by Mattriks** [comment]
Yes I would like to open a PR for this sometime (with new functions for `Gadfly.vstack`, `Gadfly.hstack` and `Gadfly.gridstack`). There is an issue here that currently `Gadfly.{hv}stack` use `Compose.{hv}stack`, but the new functions (e.g. `vstack2` above)  do not use `Compose.{hv}stack`.  I think `Compose.{hv}stack` should remain as they are, The question then is: what is the best way of incorporating the new `Gadfly._stack` functions into `Gadfly`? e.g. should they be renamed?

**Comment by bjarthur** [comment]
i'd suggest making a new function, say `figure()`, which when given a column vector of `Plot`s acts like `vstack`, given a row vector like `hstack`, and given a matrix like `gridstack`.  keyword arguments could replace `title` and specify spacing.

would be great to also be able to do insets, as well as to align plots by their axis origins.

see https://cran.r-project.org/web/packages/egg/vignettes/Ecosystem.html for how R implements all of this functionality.

my only hesitation is that this approach is not declarative.

**Comment by Mattriks** [comment]
I like the suggestion of a `figure()` function (tbd). I view it as a helper function, for users who want that.  If you want declarative graphics than that functionality is in `Compose` e.g. here is an inset plot:

```Julia
using Compose, DataFrames, Gadfly
pa = plot(XPI, x=:Year, y=:TCC, Geom.line,
    Coord.cartesian(ymax=150),
    Theme(default_color="red", panel_fill="CornSilk") )
pb = plot(XPI, x=:Year, y=:CPI, Geom.line, Coord.cartesian(ymin=10),
    Guide.ylabel("CPI (\$US '000s)"),
    Guide.title("Xmas Price Index
(https://en.wikipedia.org/wiki/Christmas_Price_Index)"),
    Guide.annotation(compose(context(0w,0h,0.75w,0.5h), render(pa))),
    Theme(default_color="green")
)
```
![xpi](https://user-images.githubusercontent.com/18226881/50052861-f4700580-017e-11e9-9c6d-3f0154927b5e.png)

**Comment by jonathanBieler** [comment]
One issue with `figure` is that it's used to open and manage figures in Matlab-like figure manager. But since it's either without argument or just an index it could be fine.

What about a `subplot` function for subplots, that work like `layer` with additional position/size arguments ?

```julia
plot(
    layer(x=x, y=y, Geom.point)
    subplot(x=z, y=w, Geom.line, position=(0w,0h,0.75w,0.5h) ),
)
```

**Comment by Mattriks** [comment]
In Gadfly Grammar of Graphics, my plot above could be done perhaps like this:
```Julia
plot( layer(x=Year, y=CPI, Geom.line, Theme(...)),
    layer(x=Year, y=TCC, Geom.inset(Geom.line, pos=(0.0w,0.0h,0.75w,0.5h)), Theme(...) ) 
)
```
similar in syntax to `Geom.subplot_grid()`.

**Comment by CNOT** [comment]
Let me put this here too. It would be really nice, if one had the option to use the same legend or title or axis labels for a stack of plots. `Geom.subplot_grid` assumes there is some sort of categorized structure to your data, so it doesn't satisfy the simple cases where you basically just want to display the same data in different scales without repeating the legend.

**Comment by Mattriks** [comment]
Here's my quick attempt at such a plot:
```julia
using Compose, DataFrames, Gadfly, RDatasets
mtcars = sort(dataset("datasets", "mtcars"), :Cyl)

myplot(D) = plot(D, x=:MPG, y=:WT, color=:Cyl, Geom.point, 
    Scale.color_discrete, Theme(key_position=:none, plot_padding=[1mm]))
Dp = by(mtcars, :Cyl, myplot) 

p1 = plot(mtcars, x=:MPG, y=:WT, color=:Cyl, Geom.point, 
    Scale.color_discrete, Theme(key_position=:none, plot_padding=[1mm]))

p2 = gridstack(reshape(Union{Plot,Compose.Context}[Dp.x1; context()], 2, 2))

gck = Guide.manual_color_key("", ["4","6","8"].*"\t\t", Scale.color_discrete().f(3))
p3 = compose(context(0.4,0.5), render(gck, Theme(), Gadfly.Aesthetics())[1].ctxs[3])

# vstack2 is in a post above
vstack2([hstack(p1, p2); p3], heights=[9,1])
```
![grid_arrange](https://user-images.githubusercontent.com/18226881/67053989-d1dbbe00-f18e-11e9-8ae8-ae0b28c8715a.png)

There are some color issues in p2, and the development of a `Geom.subplot_wrap` would help (#855).

**Comment by Mattriks** [comment]
Example 2
```julia
gck = Guide.manual_color_key("Cylinders", ["4","6","8"].*"\t\t", Scale.color_discrete().f(3))
p3 = compose(context(),(context(0.5,0), render(gck, Theme(), Gadfly.Aesthetics())[1].ctxs[1]))
p2 = gridstack(reshape(Union{Plot,Compose.Context}[Dp.x1; p3], 2, 2))
hstack(p1, p2)
```
![grid_arrange](https://user-images.githubusercontent.com/18226881/67058491-c8f3e800-f1a0-11e9-93cf-413676eb50be.png)

**Comment by CNOT** [comment]
This actually helps a lot. Thanks.
